shared/data-scripts/user-data-server.sh

```diff
@@ -15,31 +15,73 @@ sed -i "s/CONSUL_TOKEN/${nomad_consul_token_secret}/g" /etc/nomad.d/nomad.hcl
 sudo systemctl restart nomad
 
 echo "Finished server setup"
-sleep 10
-echo "ACL bootstrap begin"
-
-# Bootstrap consul ACLs
-consul acl bootstrap | grep -i secretid | awk '{print $2}' > $CONSUL_BOOTSTRAP_TOKEN
-
-if [ $? -eq 0 ]; then
-    consul acl policy create -name 'nomad-auto-join' -rules="@$ACL_DIRECTORY/consul-acl-nomad-auto-join.hcl" -token-file=$CONSUL_BOOTSTRAP_TOKEN
-
-    consul acl role create -name "nomad-auto-join" -description "Role with policies necessary for nomad servers and clients to auto-join via Consul." -policy-name "nomad-auto-join" -token-file=$CONSUL_BOOTSTRAP_TOKEN
-
-    consul acl token create -accessor=${nomad_consul_token_id} -secret=${nomad_consul_token_secret} -description "Nomad server/client auto-join token" -role-name nomad-auto-join -token-file=$CONSUL_BOOTSTRAP_TOKEN
 
-    # Wait for nomad servers to come up
-    sleep 40
-
-    # Bootstrap nomad ACLs
-    nomad acl bootstrap | grep -i secret | awk -F '=' '{print $2}' | xargs > $NOMAD_BOOTSTRAP_TOKEN
-
-    nomad acl policy apply -token $(cat $NOMAD_BOOTSTRAP_TOKEN) -description "Policy to allow reading of agents and nodes and listing and submitting jobs in all namespaces." node-read-job-submit $ACL_DIRECTORY/nomad-acl-user.hcl
-
-    nomad acl token create -token $(cat $NOMAD_BOOTSTRAP_TOKEN) -name "read-token" -policy node-read-job-submit | grep -i secret | awk -F "=" '{print $2}' | xargs > $NOMAD_USER_TOKEN
+echo "ACL bootstrap begin"
 
-    # Write user token to kv
-    consul kv put -token-file=$CONSUL_BOOTSTRAP_TOKEN nomad_user_token $(cat $NOMAD_USER_TOKEN)
-fi
+# Wait until leader has been elected and bootstrap consul ACLs
+while true; do
```

Comment:
`while true` exposes you to the risk of having the script pending forever in case the Consul dc never elects a leader. If you want to use it in a headless environment you could use something that gives you a limited amount of attempts before failing.

For instruqt I used something like this:

```bash
log "ACL Bootstrap"

for i in `seq 1 9`; do

  consul acl bootstrap --format json > ${ASSETS}/acl-token-bootstrap.json 2> /dev/null;

  excode=$?

  if [ ${excode} -eq 0 ]; then
    break;
  else
    if [ $i -eq 9 ]; then
      echo -e '\033[1m\033[31m[ERROR] \033[0m Failed to bootstrap ACL system, exiting.';
      exit 1
    else
      echo -e '\033[1m\033[33m[WARN] \033[0m ACL system not ready. Retrying...';
      sleep 5;
    fi
  fi

done

export CONSUL_HTTP_TOKEN=`cat ${ASSETS}/acl-token-bootstrap.json | jq -r ".SecretID"`
```